doc/arch/adr-006-geoprocessing-multioperation.md

```diff
@@ -0,0 +1,65 @@
+
+The above can be done either to the existing operations, and adapting their output in MMW to account for the new list return format. Previous operations can still be supported with just a singleton list as the output. Or the above can be implemented as new, "multishape" versions of the same operations.
+
+Furthermore, since a number of the geoprocessing calls invoked by MapShed query the same rasters, e.g. [n](https://github.com/WikiWatershed/model-my-watershed/blob/3ec450adc9ffb84b5483ccdc338cdeacb36f8b2b/src/mmw/mmw/settings/base.py#L463)[l](https://github.com/WikiWatershed/model-my-watershed/blob/3ec450adc9ffb84b5483ccdc338cdeacb36f8b2b/src/mmw/mmw/settings/base.py#L478)[c](https://github.com/WikiWatershed/model-my-watershed/blob/3ec450adc9ffb84b5483ccdc338cdeacb36f8b2b/src/mmw/mmw/settings/base.py#L513)[d](https://github.com/WikiWatershed/model-my-watershed/blob/3ec450adc9ffb84b5483ccdc338cdeacb36f8b2b/src/mmw/mmw/settings/base.py#L537), including them in them in the same geoprocessing call would be a speed up as well. This would require not only sending a set of shapes, but a set of operations as well, each with an `operationType` and a list of `rasters`. The output should be a list of lists of sets of results, where the first list matches the list of component shapes, and the second the list of operations.
+
+Since both the input and output of this operation can be quite different than the current implementation, it should be added as a new operation type rather than a modification to an existing one. MapShed should be updated to use this new format. The new operation should also support submitting a single shape instead of a list of shapes, i.e. the use case for a single HUC-12. In this case the output should be a singleton list of list of results.
+
+## Decision
+
+We will make add a new operation to the geoprocessing service that takes a list of shapes and a list of operations. The first shape in the list is assumed to be the parent shape that contains all the other child shapes. Using the parent shape we pull tiles corresponding to every raster in the list of operations, and collect them in a [`CompositeTile`](https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/CompositeTile.scala) per raster.
+
+Then, for each child shape, we go through the list of operations and calculate the results using the respective `CompositeTile`s from above. All the results are then collected in order and returned as a list.
+
+Once this ADR is approved, the next steps will be:
+
+  1. Define input and outputs for new geoprocessing operation
+  2. Implement new geoprocessing operation
+  3. Benchmark the new operation against current ones
+  4. Make a new release of the geoprocessing service and publish it
+  5. Update MMW to use the new geoprocessing service
+  6. Update MapShed to use the new operation
+  7. Update Subbasin code to use the new operation
+
+## Consequences
```

Comment:
It seems like an additional consequence would be our ability to create and use cached results at the huc12 level when subbasin is on.  Could you add a description of the impact on our use of caching?  I think we could either make the geop service aware of the cache instance, or check and populate cache results from celery before submitting and after getting the results back from the service.

Reply:
Hmm good point. I think we should keep the caching at the Celery level. Here's an outline of the steps I'm imagining:

1. `[django]` Receive request, start Celery job
2. `[celery]` Enumerate all the shapes and operations needed for a subbasin MapShed run
3. `[celery]` Check the cache for their keys, and make a list of all that are not available. Invoke geoprocessing.
4. `[geoprocessing]` Execute the new multi-operation geoprocessing for the ones not available
5. `[celery]` Once the results come back, cache them with the right key
6. `[celery]` Get all results out of the cache and return

I'll add this to the Consequences section.

Reply:
Added in 3088461e.